### PR TITLE
feat: support all deploy workflows

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?

Feature branch for supporting all deploy workflows

### What issues does this PR fix or reference?
[skip-validate-pr]

### CODE REVIEW

File: src/utils/output.ts [hunk 1]
```diff
@@ -19,19 +19,23 @@ import {
...
-import { API, TestLevel } from './types';
+import { Messages, NamedPackageDir, SfProject } from '@salesforce/core';
+import { StandardColors } from '@salesforce/sf-plugins-core';
+import { API, AsyncDeployResultJson, DeployResultJson, RetrieveResultJson, TestLevel, Verbosity } from './types';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'deploy.async');
 
 function info(message: string): string {
   return blue(bold(message));
```
Comment (peternhale) [inline src/utils/output.ts:30]: This deviates from StandardColors.info.

File: src/utils/output.ts [hunk 2]
```diff
@@ -86,189 +93,347 @@ export function toArray<T>(entryOrArray: T | T[] | undefined): T[] {
...
     CliUx.ux.log();
+    CliUx.ux.log(info('Test Results Summary'));
+    const passing = this.result.response.numberTestsCompleted ?? 0;
+    const failing = this.result.response.numberTestErrors ?? 0;
+    const total = this.result.response.numberTestsTotal ?? 0;
+    const time = this.result.response.details.runTestResult.totalTime ?? 0;
+    CliUx.ux.log(`Passing: ${passing}`);
+    CliUx.ux.log(`Failing: ${failing}`);
+    CliUx.ux.log(`Total: ${total}`);
+    if (time) CliUx.ux.log(`Time: ${time}`);
```
Comment (peternhale) [inline src/utils/output.ts:235]: Can we get these colored appropriately

File: src/utils/output.ts [hunk 3]
```diff
@@ -86,189 +93,347 @@ export function toArray<T>(entryOrArray: T | T[] | undefined): T[] {
...
+
+  private displayVerboseTestSuccesses(): void {
+    const successes = toArray(this.result.response.details.runTestResult?.successes);
+    if (successes.length > 0) {
+      const testSuccesses = sortTestResults(successes) as Successes[];
+      CliUx.ux.log();
+      CliUx.ux.log(success(`Test Success [${successes.length}]`));
+      for (const test of testSuccesses) {
+        const testName = underline(`${test.name}.${test.methodName}`);
+        CliUx.ux.log(`${check} ${testName}`);
```
Comment (peternhale) [inline src/utils/output.ts:246]: I know check is colored using color success, what about the name of the test?

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/utils/output.ts (hunk 2 above)
Comment (mdonnalley) [reply src/utils/output.ts:235]: @peternhale what color is appropriate for this?

File: src/utils/output.ts (hunk 3 above)
Comment (mdonnalley) [reply src/utils/output.ts:246]: It's also the success color

File: src/utils/output.ts (hunk 1 above)
Comment (mdonnalley) [reply src/utils/output.ts:30]: @peternhale This is only being used for the titles on tables. I'll rename the function appropriately and keep it blue - unless you think there's a better color for that.

File: src/utils/errorCodes.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import { RequestStatus } from '@salesforce/source-deploy-retrieve';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-deploy-retrieve', 'errorCodes');
+
+export const DEPLOY_STATUS_CODES = new Map<RequestStatus, number>([
+  [RequestStatus.Succeeded, 0],
+  [RequestStatus.Canceled, 1],
+  [RequestStatus.Failed, 1],
+  [RequestStatus.SucceededPartial, 68],
```
[the remark this answers is not in the document]
Comment (mdonnalley) [reply src/utils/errorCodes.ts:18]: This is taken directly from plugin-source

File: src/utils/output.ts (hunk 2 above)
Comment (peternhale) [reply src/utils/output.ts:235]: I believe it should be info. But I am thinking that maybe Matt should have a look at all of the output and make recommendations that best suit the deploy commands.

File: src/commands/deploy/metadata.ts [hunk 5]
```diff
@@ -97,13 +111,15 @@ export default class DeployMetadata extends SfCommand<DeployMetadataResult> {
       defaultValue: 33,
```
[the remark this answers is not in the document]
Comment (peternhale) [reply src/commands/deploy/metadata.ts:111]: Not really, just curious.

**Review by mattgraham — APPROVED** [review]
:shipit: We walked through this earlier today. Looks good to ship.

**Review by peternhale — APPROVED** [review]

**Pull request merged**